src/lib/hardware/pci.c

```diff
@@ -54,9 +54,14 @@ uint32_t volatile *map_pci_resource(int fd)
   }
 }
 
-void close_pci_resource(int fd)
+void close_pci_resource(int fd, uint32_t volatile *addr)
```

Comment:
I have been clarifying `volatile` and related issues with Mike Pall on the LuaJIT list.

Conclusions:

There *is* a risk that the JIT optimizations will eliminate or reorder loads and stores.

`volatile` is not the solution: LuaJIT [ignores](http://www.freelists.org/post/luajit/Compiler-loadstore-barrier-volatile-pointer-barriers-in-general,8) that. Probably we should not use it: I see a risk that we will expect it to behave in a way that it does not.

[Compiler barriers](http://www.freelists.org/post/luajit/Compiler-loadstore-barrier-volatile-pointer-barriers-in-general,5) are a solution. For example, immediately before reading or writing a memory mapped register we could always call a `lib.c` function:

```
void compiler_barrier() {}
```

and LuaJIT will flush loads/stores from registers to memory before making an FFI call.